test/distributed/pipeline/sync/test_pipe.py

```diff
@@ -381,22 +381,124 @@ class NonTensorTuple(nn.Module):
         def forward(self, x):
             return (x, "hello")
 
+    class NonTensorArgs(nn.Module):
+        def forward(self, x: str, y: bool):
+            return x, y
+
     model = nn.Sequential(NonTensorTuple())
     model = Pipe(model)
     x = torch.rand(1)
 
-    # TypeError: CheckpointBackward.forward: expected Variable (got str) for return value 1
-    with pytest.raises(TypeError):
-        model(x)
-
-    # TypeError: expected Tensor to scatter, but got str
     with pytest.raises(TypeError):
         model((x, "hello"))
 
-    # TypeError: expected Tensor to scatter, but got str
     with pytest.raises(TypeError):
         model([x, "hello"])
 
+    model = nn.Sequential(NonTensorArgs())
+    model = Pipe(model)
+
+    with pytest.raises(TypeError):
+        model("hello", True)
+
+
+@pytest.mark.parametrize("checkpoint", ["never", "always", "except_last"])
+def test_valid_non_tensor(checkpoint, setup_rpc):
+    class NonTensor1(nn.Module):
+        def forward(self, a: int, b: Tensor, c: bool, d: Tensor):
+            res = b + a if c else b * a
+            if d is not None:
+                res += d
+            return res, c, a, b, "hello", d
+
+    class NonTensor2(nn.Module):
+        def forward(self, a: Tensor, b: bool, c: int, d: Tensor, e: str, f: Tensor):
+            res = a * c if b else a + c
+            res += d
+            return c, res, a, d + f if f is not None else d, b, e, f
+
+    model = Pipe(nn.Sequential(NonTensor1(), NonTensor2()), chunks=5, checkpoint=checkpoint)
+    a = random.randint(0, 10)
+    b = torch.rand(10, 10)
+    c = random.randint(0, 1) == 0
+    d = torch.rand(10, 10)
+    res = model(a, b, c, d).local_value()
+    assert 5, len(res)
+    assert [a] * 5 == res[0]
+    if c:
+        assert torch.allclose(((b + a + d) * a) + b, res[1])
+        assert torch.allclose(b + a + d, res[2])
+    else:
+        assert torch.allclose(((b * a) + d + a) + b, res[1])
+        assert torch.allclose(b * a + d, res[2])
+    assert torch.allclose(b + d, res[3])
+    assert [c] * 5 == res[4]
+    assert ["hello"] * 5 == res[5]
+    assert torch.allclose(d, res[6])
+
+    # Test one of the tensors can be None
+    res = model(a, b, c, None).local_value()
```

Comment:
Can you also make `b` as a tensor? This way you also test 1) multiple tensor args; 2) a tensor arg does not have to be the last arg.